homeassistant/helpers/script.py

```diff
@@ -309,173 +222,288 @@ def async_log_exception(self, logger, message_base, exception):
-                    template.render_complex(action[CONF_EVENT_DATA_TEMPLATE], variables)
+                    template.render_complex(
+                        self._action[CONF_EVENT_DATA_TEMPLATE], self._variables
+                    )
                 )
             except exceptions.TemplateError as ex:
-                _LOGGER.error("Error rendering event data template: %s", ex)
+                self._log(
+                    "Error rendering event data template: %s", ex, level=logging.ERROR
+                )
 
-        self.hass.bus.async_fire(action[CONF_EVENT], event_data, context=context)
+        self._hass.bus.async_fire(
+            self._action[CONF_EVENT], event_data, context=self._context
+        )
 
-    async def _async_check_condition(self, action, variables, context):
+    async def _async_condition_step(self):
         """Test if condition is matching."""
-        config_cache_key = frozenset((k, str(v)) for k, v in action.items())
+        config_cache_key = frozenset((k, str(v)) for k, v in self._action.items())
         config = self._config_cache.get(config_cache_key)
         if not config:
-            config = await condition.async_from_config(self.hass, action, False)
+            config = await condition.async_from_config(self._hass, self._action, False)
             self._config_cache[config_cache_key] = config
 
-        self.last_action = action.get(CONF_ALIAS, action[CONF_CONDITION])
-        check = config(self.hass, variables)
-        self._log(f"Test condition {self.last_action}: {check}")
-
+        self.last_action = self._action.get(CONF_ALIAS, self._action[CONF_CONDITION])
+        check = config(self._hass, self._variables)
+        self._log("Test condition %s: %s", self.last_action, check)
         if not check:
             raise _StopScript
 
-    def _async_set_timeout(self, action, variables, context, continue_on_timeout):
-        """Schedule a timeout to abort or continue script."""
-        timeout = action[CONF_TIMEOUT]
-        unsub = None
+    def _log(self, msg, *args, level=logging.INFO):
+        self._script._log(msg, *args, level=level)  # pylint: disable=protected-access
+
+    def _raise(self, msg, *args, exception=None):
+        # pylint: disable=protected-access
+        self._script._raise(msg, *args, exception=exception)
+
+
+class _ScriptRun(_ScriptRunBase):
+    """Manage asynchronous run of Script sequence."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        script: "Script",
+        variables: Optional[Sequence],
+        context: Optional[Context],
+    ) -> None:
+        super().__init__(hass, script, variables, context)
+        self.task: Optional[asyncio.Task] = None
+
+    async def async_run(self) -> None:
+        """Run script."""
+        self._log("Running script")
+        await self._async_run(0)
+
+    def stop(self) -> None:
+        """Stop script run."""
+        cast(asyncio.Task, self.task).cancel()
```

Comment:
This can be `None`, seems dangerous to just cast it.

Reply:
I'm also relatively new to typing in Python. I can tell you I had to make quite a few changes to make pylint, mypy, etc. happy. I actually think it makes the code _less_ clear. But that's neither here nor there. The main point is, no, it can't be None. For the non-legacy _ScriptRun, there will _always_ be a Task in `self.task` that `async_run`, etc. run in. That's part of the design, so that things can get cleaned up if told to stop.